### PR TITLE
Fixed Full Scan graph crash on mixed-polarity transition lists (#4240)

### LINKED ISSUE #4240: GraphFullScan: AssumptionException (polarity mismatch) crashes Full Scan viewer for mixed-polarity / negative-mode IMS documents
## Problem

The Full Scan graph viewer crashes with an unhandled `AssumptionException` when `CreateSingleScan` iterates a transition whose polarity does not match the displayed scan's polarity. The polarity assertion fires *before* the per-transition source/m-z skip check, so any opposite-polarity transition present in `ScanProvider.Transitions` brings down the app. It is reached via the timer-driven `SkylineWindow.UpdateGraphPanes` while the Full Scan graph is open.

## Exception Report

- **Fingerprint:** `eddb8a1125c183f7`
- **Exception ID (skyline.ms rowId):** 74770
- **Link:** https://skyline.ms/home/issues/exceptions/announcements-thread.view?rowId=74770
- **Version:** 26.1.0.057-c07debd50 (64-bit)
- **Reports:** 1 report from 1 user (2026-05-20)
- **Error message:** *(empty — `Assume.IsTrue` was called with no message argument)*

## Stack Trace

```
pwiz.Common.SystemUtil.AssumptionException
   at pwiz.Common.SystemUtil.Assume.Fail(String error) in Assume.cs:line 147
   at pwiz.Skyline.Controls.Graphs.GraphFullScan.CreateSingleScan(Double[]& massErrors) in GraphFullScan.cs:line 976
   at pwiz.Skyline.Controls.Graphs.GraphFullScan.CreateGraph() in GraphFullScan.cs:line 382
   at pwiz.Skyline.Controls.Graphs.GraphFullScan.UpdateUI(Boolean selectionChanged) in GraphFullScan.cs:line 1355
   at pwiz.Skyline.SkylineWindow.UpdateGraphPanes(Object sender, EventArgs e) in SkylineGraphs.cs:line 881
   at System.Windows.Forms.Timer.OnTick(EventArgs e)
   ...
   at pwiz.Skyline.Program.Main(String[] args) in Program.cs:line 354
```

## Root Cause

In `GraphFullScan` (release: `CreateSingleScan`; master: `CreateSingleScanInPane`), the loop over `ScanProvider.Transitions` asserts polarity equality before the skip check:

```csharp
// Polarity should match, because these are the spectra used for extraction
Assume.IsTrue(transition.PrecursorMz.IsNegative == negativeScan);
if (transition.Source != _msDataFileScanHelper.Source ||
    !transition.MatchMz(mz))
    continue;
```

The assertion has been present since 2016 (commit `72fa33f37`, the SignedMz / negative-ion-mode support change). It assumes the transition list never contains a transition of the opposite polarity to the scan being drawn. That assumption is violated in at least two scenarios.

## Likely Triggers

1. **Mixed-polarity document** — both positive and negative precursors present. The displayed scan is a single polarity, so the opposite-polarity transitions trip the assertion even though they would simply be skipped by the `Source`/`MatchMz` check below.
2. **Ion-mobility 2-D path** — `negativeScan = fullScans.Any() && fullScans.First().NegativeCharge`. If `GetFilteredScans` returns nothing (e.g. the IM filter window excludes all scans), `negativeScan` silently defaults to `false`; a negative-mode transition then fails the assertion.

The original report did not capture the document, so the exact trigger is unconfirmed — both should be covered by the fix.

## Proposed Fix

Fold the polarity check into the existing skip condition instead of asserting, so the viewer ignores non-matching-polarity transitions (which the comment says is the intent):

```csharp
if (transition.PrecursorMz.IsNegative != negativeScan ||
    transition.Source != _msDataFileScanHelper.Source ||
    !transition.MatchMz(mz))
    continue;
```

## Files to Modify

- `pwiz_tools/Skyline/Controls/Graphs/GraphFullScan.cs` — the `Assume.IsTrue(transition.PrecursorMz.IsNegative == negativeScan)` line in the single-scan point-assignment loop.

## Status / Branches

- Present in **master** — must be fixed there first.
- Present in **skyline_26_1** release branch (origin of this crash). Needs cherry-pick — see label.

## Test

Add a functional test that opens the Full Scan graph for a mixed-polarity (or negative-mode IMS) small-molecule document and clicks a chromatogram point (`ClickChromatogram`). Candidate data: `NegativeIonChromatogramsTest.zip` (mixed polarity small molecules — needs full-scan MS1 settings added), or extend `FullScanGraphTest`'s IMS data with mixed-polarity transitions. Without the fix the test crashes in `CreateSingleScan`.

## Getting Started

Use /pw-startissue <number> to begin work.

---
Created from conversation context (exception triage). Source: skyline.ms exception rowId 74770.

### PR DESCRIPTION
**Pull request by bspratt**
## Summary

* `GraphFullScan.CreateSingleScan` now skips transitions whose polarity differs from the displayed scan instead of asserting — fixes the `AssumptionException` in exception report #74770.
* Added `MixedPolarityFullScanTest`: a regression test that builds a `ScanProvider` whose transition list spans both polarities and opens the Full Scan on a positive scan. A plain UI click can't reproduce it (opposite-polarity precursors always render in separate panes), so the test constructs the condition directly via `ShowGraphFullScan`.
* Added a scan **Polarity** row to the Full Scan Acquisition property sheet (useful for mixed-polarity data), and made "True"/"False" localizable for "Is Centroided" display. This is the only user-facing change.
 
image

Fixes #4240

## Test plan

- [x] TestMixedPolarityFullScan — fails with `AssumptionException` (GraphFullScan.cs:1951) without the fix, passes with it
- [x] TestFullScanGraph — passes with the new Polarity property

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

File: pwiz_tools/Skyline/TestFunctional/MixedPolarityFullScanTest.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            var csv = TextUtil.LineSeparate(
+                "MixedPol,TestMol,C20H30O2,1",
+                "MixedPol,TestMol,C20H30O2,-1").Replace(',', TextUtil.CsvSeparator);
+            var insertDlg = ShowDialog<InsertTransitionListDlg>(SkylineWindow.ShowPasteTransitionListDlg);
+            var colDlg = ShowDialog<ImportTransitionListColumnSelectDlg>(() => insertDlg.TransitionListText = csv);
+            RunUI(() =>
+            {
+                colDlg.radioMolecule.PerformClick();
+                colDlg.SetSelectedColumnTypes(
+                    Resources.ImportTransitionListColumnSelectDlg_ComboChanged_Molecule_List_Name,
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/TestFunctional/MixedPolarityFullScanTest.cs:89]: Addressed in 0c79eb9ce: switched to `RunDlg<ImportTransitionListColumnSelectDlg>`, which waits for the dialog to be shown before its controls are touched (the interaction runs once the dialog is up). Thanks!

File: pwiz_tools/Skyline/Model/FullScanProperties.cs [hunk 2]
```diff
@@ -117,7 +117,11 @@ public static FullScanProperties CreateProperties(MsDataSpectrum spectrum)
             }
             res.Instrument.InstrumentSerialNumber = spectrum.InstrumentSerialNumber;
 
-            res.IsCentroided = spectrum.Centroided.ToString(CultureInfo.CurrentCulture);
+            res.IsCentroided = spectrum.Centroided ? FullScanPropertiesRes.True : FullScanPropertiesRes.False;
+
+            res.Polarity = spectrum.NegativeCharge
+                ? FullScanPropertiesRes.Polarity_Negative
+                : FullScanPropertiesRes.Polarity_Positive;
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/Model/FullScanProperties.cs:124]: Good catch — fixed in 756174ebc. Added `{"Polarity", FullScanPropertiesRes.Polarity_Positive}` to `Ms1FullScanFilteringTutorial.TestFullScanProperties` (the MS1 data is positive-mode) and switched `IsCentroided` to `FullScanPropertiesRes.False` for localization safety. `TestMs1Tutorial` passes.

File: pwiz_tools/Skyline/TestFunctional/FullScanGraphTest.cs [hunk 3]
```diff
@@ -81,6 +81,7 @@ protected override void DoTest()
                 {"DataPoints",105373.ToString(@"N0", CultureInfo.CurrentCulture)},
                 {"MzCount",45751.ToString(@"N0", CultureInfo.CurrentCulture)},
                 {"IsCentroided","False"},
+                {"Polarity","positive"},
                 {"idotp",0.84.ToString(CultureInfo.CurrentCulture)}
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/TestFunctional/FullScanGraphTest.cs:85]: Addressed in an earlier review-response commit — the expected value now uses `FullScanPropertiesRes.Polarity_Positive` (resource-driven, so it stays translation-proof).

File: pwiz_tools/Skyline/TestFunctional/FullScanGraphTest.cs [hunk 4]
```diff
@@ -101,6 +102,7 @@ protected override void DoTest()
                 {"DataPoints",67630.ToString(@"N0", CultureInfo.CurrentCulture)},
                 {"MzCount",31378.ToString(@"N0", CultureInfo.CurrentCulture)},
                 {"IsCentroided","False"},
+                {"Polarity","positive"},
                 { "dotp", 0.81.ToString(CultureInfo.CurrentCulture) }
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/TestFunctional/FullScanGraphTest.cs:106]: Addressed in an earlier review-response commit — the expected value now uses `FullScanPropertiesRes.Polarity_Positive` (resource-driven, so it stays translation-proof).

File: pwiz_tools/Skyline/TestFunctional/FullScanGraphTest.cs [hunk 5]
```diff
@@ -132,6 +134,7 @@ protected override void DoTest()
                 { "DataPoints", 60587.ToString(@"N0", CultureInfo.CurrentCulture) },
                 { "MzCount", 29876.ToString(@"N0", CultureInfo.CurrentCulture) },
                 { "IsCentroided", "False" },
+                { "Polarity", "positive" },
                 { "dotp", 0.51.ToString(CultureInfo.CurrentCulture) }
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/TestFunctional/FullScanGraphTest.cs:138]: Addressed in an earlier review-response commit — the expected value now uses `FullScanPropertiesRes.Polarity_Positive` (resource-driven, so it stays translation-proof).

File: pwiz_tools/Skyline/Model/FullScanProperties.cs [hunk 6]
```diff
@@ -119,6 +119,9 @@ public static FullScanProperties CreateProperties(MsDataSpectrum spectrum)
 
             res.IsCentroided = spectrum.Centroided.ToString(CultureInfo.CurrentCulture);
 
+            res.Polarity = FullScanPropertiesRes.ResourceManager.GetString(
+                spectrum.NegativeCharge ? @"Polarity_negative" : @"Polarity_positive");
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/Model/FullScanProperties.cs]: Addressed in an earlier review-response commit — `Polarity` and `IsCentroided` now use the strongly-typed `FullScanPropertiesRes` accessors (`Polarity_Positive`/`Polarity_Negative`, `True`/`False`) rather than string-literal `ResourceManager.GetString` keys.

File: pwiz_tools/Skyline/Controls/Graphs/GraphFullScan.cs [hunk 7]
```diff
@@ -1947,9 +1947,12 @@ private void CreateSingleScanInPane(MSGraphPane targetPane, out double[] massErr
...
-                    Assume.IsTrue(transition.PrecursorMz.IsNegative == negativeScan);
-                    if (transition.Source != _msDataFileScanHelper.Source ||
+                    // Normally these are the spectra used for extraction, so polarity matches. But the
+                    // transition list can include co-displayed precursors of the opposite polarity (e.g. a
+                    // molecule with both [M+H]+ and [M-H]- precursors shown together in one chromatogram
+                    // pane), so just skip any whose polarity does not match this scan rather than asserting.
+                    if (transition.PrecursorMz.IsNegative != negativeScan ||
+                        transition.Source != _msDataFileScanHelper.Source ||
                         !transition.MatchMz(mz))
                         continue;
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/Controls/Graphs/GraphFullScan.cs]: Addressed in bfce1d8ad. The crash fix only guarded the point-assignment loop; the curve-creation, transition-label, and mass-error loops still filtered by Source alone. Factored the test into a `TransitionAppliesToScan` helper (Source + polarity) and applied it at all four enumeration sites, so an opposite-polarity precursor in a mixed list is ignored consistently — no empty curve, stray label, or bogus mass error. Also strengthened TestMixedPolarityFullScan to assert the off-polarity transition renders no curve. Good catch, thanks!

File: pwiz_tools/Skyline/Controls/Graphs/GraphFullScan.cs [hunk 8]
```diff
@@ -1292,12 +1292,12 @@ private void CreateGraph()
 
                 GetRankedSpectrum();
                 double[] massErrors;
-                CreateSingleScan(out massErrors);
+                CreateSingleScan(out massErrors, out var negativeScan);
 
                 PopulateProperties();
                 AddExtractionBoxes(GraphPane);
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/Controls/Graphs/GraphFullScan.cs:1298]: Fixed in b4725be69. AddExtractionBoxes now takes the displayed scan polarity and filters with the same `TransitionAppliesToScan` helper, so the opposite-polarity precursor no longer draws a stray extraction box. Updated all six call sites (passing `negativeScan` where in scope, `IsDisplayedScanNegative()` on the heatmap paths) and extended TestMixedPolarityFullScan to assert a single extraction box. Thanks again!

File: pwiz_tools/Skyline/Controls/Graphs/GraphFullScan.cs [hunk 9]
```diff
@@ -1204,9 +1204,9 @@ private void CreateGraph()
                     CreateMobilogram();
 
                     PopulateProperties();
-                    AddExtractionBoxes(_heatMapPane);
+                    AddExtractionBoxes(_heatMapPane, IsDisplayedScanNegative());
                     if (!_showIonSeriesAnnotations)
-                        AddTransitionLabels(_heatMapPane, null);
+                        AddTransitionLabels(_heatMapPane, null, IsDisplayedScanNegative());
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/Controls/Graphs/GraphFullScan.cs]: Fixed in 237a76617 — each heatmap branch now computes `var negativeScan = IsDisplayedScanNegative();` once and passes it to both AddExtractionBoxes and AddTransitionLabels, so GetFilteredScans runs once per render. Thanks!

File: pwiz_tools/Skyline/Controls/Graphs/GraphFullScan.cs [hunk 10]
```diff
@@ -1234,9 +1234,9 @@ private void CreateGraph()
                     CreateIonMobilityHeatmap();
 
                     PopulateProperties();
-                    AddExtractionBoxes(_heatMapPane);
+                    AddExtractionBoxes(_heatMapPane, IsDisplayedScanNegative());
                     if (!_showIonSeriesAnnotations)
-                        AddTransitionLabels(_heatMapPane, null);
+                        AddTransitionLabels(_heatMapPane, null, IsDisplayedScanNegative());
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/Controls/Graphs/GraphFullScan.cs]: Fixed in 237a76617 — each heatmap branch now computes `var negativeScan = IsDisplayedScanNegative();` once and passes it to both AddExtractionBoxes and AddTransitionLabels, so GetFilteredScans runs once per render. Thanks!

**Comment by rita-gwen** [comment]
The changes look good to me. I am not sure why there would be ions of the wrong polarity in a spectrum. Is it even physically possible? For Orbitrap detectors it might be, i guess. In any case, should we just ignore such cases, or show a non-intrusive warning to the user, since it can be a data quality issue.

**Comment by bspratt** [comment]
> The changes look good to me. I am not sure why there would be ions of the wrong polarity in a spectrum. Is it even physically possible? For Orbitrap detectors it might be, i guess. In any case, should we just ignore such cases, or show a non-intrusive warning to the user, since it can be a data quality issue.

It's not really wrong polarity ions in a spectrum, it's that an entirely wrong polarity spectrum is being looked at in a polarity-switching file where both + and - scans are present
